### ISSUE TITLE
OpenWebAuth login on remote server fails if Friendica server does not have a contact

### ISSUE THREAD

**Issue by abanink** [friendica/friendica#13173]
- [X] I have searched open and closed issues for duplicates

### Bug Description

When trying to perform an OpenWebAuth login on e.g. a Hubzilla server through a Friendica login, the login fails if the Friendica server does not have a contact on the Hubzilla server. The OWA login is not even started because the Friendica server bails out before.

### Steps to Reproduce

1.  Log in on Friendica server
2.  On a Hubzilla server, go to remote authentication on /rmagic
3.  Enter the friendica handle and press 'Authenticate'

Actual Result:

The Hubzilla server returns to the login page, no user is logged in

Expected Result:

The Friendica user should be logged in on the Hubzilla server as visitor

### Screenshots

When the OWA-based login works, then the Hubzilla avatar in the left upper corner changes to the Friendica avatar. If not, it doesn't.

### Platform Info

Friendica Version: Friendica 'Giant Rhubarb' 2023.09-dev - 1518

### Troubleshooting

I see these Friendica log messages:
```
{"file":"Contact.php","line":1335,"function":"getIdForURL","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
2023-05-26T05:52:57Z index [WARNING]: E_WARNING: Undefined array key "gsid" {"code":2,"message":"Undefined array key \"gsid\"","file":"/srv/friendica/src/Module/Magic.php","line":113} - {"file":"Magic.php","line":113,"function":"rawContent","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
2023-05-26T05:52:57Z index [WARNING]: E_WARNING: Undefined array key "gsid" {"code":2,"message":"Undefined array key \"gsid\"","file":"/srv/friendica/src/Module/Magic.php","line":115} - {"file":"Magic.php","line":115,"function":"rawContent","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
2023-05-26T05:52:57Z index [NOTICE]: Server not found, redirecting to destination. {"gsid":null,"dest":"https://hzsandbox2.eenoog.org/"} - {"file":"Magic.php","line":115,"function":"rawContent","request-id":"64704938c477e","uid":"ff1c29","process_id":47915}
```

I think this is due to this code in src/Module/Magic.php:

		if ($addr ?: $dest) {
			$contact = Contact::getByURL($addr ?: $dest);
		}

...

		$gserver = $this->dba->selectFirst('gserver', ['url'], ['id' => $contact['gsid']]);
		if (empty($gserver)) {
			$this->logger->notice('Server not found, redirecting to destination.', ['gsid' => $contact['gsid'], 'dest' => $dest]);
			System::externalRedirect($dest);
		}

Indeed, it is possible to reach this code if the contact is empty but in case of OWA, that's fine: it is not required for OWA that the Friendica server knows an ActivityPub actor on Hubzilla.

In my server, I have these entries in the 'gserver' table:

> MariaDB [friendica]> select id,url from gserver;
> +----+-------------------------------+
> | id | url                           |
> +----+-------------------------------+
> |  1 | https://friendica.eenoog.org  |
> |  2 | https://hub.eenoog.org        |
> |  3 | https://stroom.eenoog.org     |
> |  4 | https://hzsandbox2.eenoog.org |
> |  5 | http://f.lapo.it              |
> |  6 | https://sandpix.eenoog.org    |
> |  7 | https://fd.winklerfamilie.eu  |
> |  8 | https://hub.somaton.com       |
> +----+-------------------------------+

In my case, the Hubzilla server is 'hzsandbox2.eenoog.org', and I have no contact on that server:
> MariaDB [friendica]> select id,uid,nick,baseurl from contact;
> +----+-----+-----------+------------------------------+
> | id | uid | nick      | baseurl                      |
> +----+-----+-----------+------------------------------+
> |  0 |   0 |           |                              |
> |  2 |   2 | pascal    |                              |
> |  3 |   0 | pascal    | https://friendica.eenoog.org |
> |  4 |   0 | friendica | https://friendica.eenoog.org |
> |  6 |   0 | sys       | http://stroom.eenoog.org     |
> |  7 |   0 | pascal    | http://stroom.eenoog.org     |
> |  8 |   2 | pascal    | http://stroom.eenoog.org     |
> |  9 |   0 | pascal    | https://stroom.eenoog.org    |
> | 10 |   0 | friendica | https://f.lapo.it            |
> | 11 |   0 | abanink   | http://sandpix.eenoog.org    |
> | 12 |   0 | friendica | https://fd.winklerfamilie.eu |
> +----+-----+-----------+------------------------------+

which explains the behaviour. 
In my view, it's a bug but I'm not enough known with Friendica internals to propose a fix.

**Comment by MrPetovan** [comment]
Changes have been added to the `src/Module/Magic.php` file since you filed this issue on the `develop` branch, are you still able to reproduce it?

**Comment by abanink** [comment]
Updated to the friendica 'develop' branch today.

> $ git log | head -1
> commit f1da323b071a0dd20680302cbce26188119f9604
 
Reproducing gives this log:

```
2023-07-26T06:58:00Z app [INFO]: Probing done {"uri":"https://hzsandbox2.eenoog.org/","network":"unkn"} - {"file":"Probe.php","line":796,"function":"detect","request-id":"64c0c3f82c51c","uid":"e48290","process_id":2208}
2023-07-26T06:58:00Z app [NOTICE]: No valid network found {"url":"https://hzsandbox2.eenoog.org/","uid":0,"default":[],"update":null,"callstack":"Contact::getByURL (91), Magic::rawContent (238), BaseModule::run (707), App::runFrontend (46)"} - {"file":"Contact.php","line":1344,"function":"getIdForURL","request-id":"64c0c3f82c51c","uid":"e48290","process_id":2208}
2023-07-26T06:58:00Z app [NOTICE]: Not logged in or not OWA, redirecting to destination. {"uid":false,"owa":1,"dest":"https://hzsandbox2.eenoog.org/"} - {"file":"Magic.php","line":110,"function":"rawContent","request-id":"64c0c3f82c51c","uid":"e48290","process_id":2208}
```
So it's not working as before for a Friendica user.
For a (streams) user, this is still working.

**Comment by MrPetovan** [comment]
Thanks for the quick follow-up, we do expect the target URL to be a profile URL. We could probably make it work for a base server URL as well. Can I use the same Hubzilla server to conduct my tests?

**Comment by abanink** [comment]
You can but the hubzilla server hzsandbox2.eenoog.org is not always reachable.

**Comment by MrPetovan** [comment]
I'm able to log in with my development node on the provided Hubzilla server, notice my profile picture in the top left:
![image](https://github.com/friendica/friendica/assets/925415/2e5236fc-8d11-4d41-909e-325d8f93cf68)

> Not logged in or not OWA, redirecting to destination.

You need to be logged in your local Friendica node for the remote authentication to work.

**Comment by MrPetovan** [comment]
Apparently fixed by #13226
